04-cs-fundamentals/internship/binary-search-trees.md

```diff
@@ -1,23 +1,253 @@
+
+![Binary Search Tree Example](images/bst1.png)
+
+The Node class for a Binary Search Tree (BST) could look like this:
+
+```ruby
+class TreeNode
+  attr_reader :data
+  attr_accessor :left, :right
+
+  def initialize(data)
+    @data = data
+    @left = @right = nil
+  end
+end
+```
+
+These nodes are arranged with one node known as the **root** which is where the tree begins.  New nodes as they are added are placed to the left, if they are less than the current node, and to the right if they are greater than the current node.  This is a natually recursive process.
+
+```
+Method add:
+  if the root is nil set the root to be a new node with the given value and return the node.
+
+  Otherwise 
+    if the value is less than or equal to the current root's value, make root's left be the result of calling add on root's left.
+    otherwise make root's right be the result of calling add on root's right.
+```
+
+You can observe this in the [Binary Tree Visualizer](http://btv.melezinek.cz/binary-search-tree.html)
+
+
+## The Tree Class
+
+You can build a `Tree` class in a manner similar to how we created a `LinkedList` class.
+
+```ruby
+class Tree
+  attr_reader :root
+
+  def initialize
+    @root = nil
+  end
+
+  # Tree methods go here...
+end
+```
+
+## Finding a Value
+
+You can try to search to find a value in a Binary Search Tree Like this:
+
+```
+If root is nil return nil
+If the root equals the value being searched for return the root's data
+
+If the value is less than the root return search on root's left side
+If the value is greater than the root return search on root's right side
+```
+
+**Exercise** 
+
+Try this out on the [Binary Tree Visualizer](http://btv.melezinek.cz/binary-search-tree.html).
+
+**Question**:  What's the worst-case for finding a value in the tree?
+
+**Question**:  What's the best-case for finding a value in the tree?
+
+**Question**:  What would you guess is the average case for finding a node in a tree?
+
+We will answer these questions in a bit.
+
+## Balanced Trees & Unbalanced Trees
+
+A tree is considered **balanced** if the height of it's left and right subtree differ by at most 1.  In this way the nodes in the tree must be spread fairly evenly.
+
+This is an example of a balanced tree.
+
+![balanced bst](images/balanced-bst.png)
+
+On the other hand this is an unbalanced tree.
+
+![unbalanced bst](images/unbalanced-bst.png)
+
+<details>
+  <summary>How many nodes do you need to examine to find 5, in the 1st example.  How many in the 2nd?</summary>
+
+  In the 1st example you look at 25, then 10, and then 5.  In the 2nd you look at 100, then 50, then 25, then 15, then 10 and then 5.  
+
+  So the 1st requires 3 lookups while the 2nd requires 6.
+</details>
+
+<br />
+
+<details>
+  <summary>Based on your expert knowledge of Big-O.  What's the time complexity to look up a node in a balanced tree?</summary>
+
+  O(log n)
```

Comment:
Are they expected to be able to do this math independently? I suspect that many would struggle with this unless given more hints.

A good intermediate step might be, "What is the maximum height of a balanced / unbalanced tree", and then walk through the reasoning that arrives at `O(log n)` and `O(n)`.